torch/utils/_pytree.py

```diff
@@ -470,75 +402,102 @@ def _broadcast_to_and_flatten(pytree: PyTree, spec: TreeSpec) -> Optional[List[A
     return result
 
 
+"""
+TreeSpecSchema is the schema used to serialize the TreeSpec
+It contains the following fields:
+- type: A string name of the type. null for the case of a LeafSpec.
+- context: Any format which is json dumpable
+- children_spec: A list of children serialized specs.
+"""
+@dataclasses.dataclass
+class TreeSpecSchema:
+    type: Optional[str]
+    context: DumpableContext
+    children_spec: List['TreeSpecSchema']
+
+
+def treespec_dumps(treespec: TreeSpec, protocol: Optional[int] = None) -> str:
+    def treespec_to_json(spec: TreeSpec) -> TreeSpecSchema:
+        if isinstance(spec, LeafSpec):
+            return TreeSpecSchema(None, None, [])
+
+        if spec.type not in SUPPORTED_SERIALIZED_TYPES:
+            raise NotImplementedError(f"Serializing {spec.type} in pytree is not registered.")
+
+        serialize_node_def = SUPPORTED_SERIALIZED_TYPES[spec.type]
+
+        serialized_type = serialize_node_def.serialized_type
+
+        if serialize_node_def.getstate is None:
+            try:
+                serialized_context = json.dumps(spec.context)
+            except TypeError:
+                raise TypeError(
+                    "Unable to serialize context. "
+                    "Please make the context json dump-able, or register a "
+                    "custom serializer using _register_treespec_serializer."
+                )
+        else:
+            serialized_context = serialize_node_def.getstate(spec.context)
+
+        child_schemas = [treespec_to_json(child) for child in spec.children_specs]
+
+        return TreeSpecSchema(serialized_type, serialized_context, child_schemas)
+
+    if protocol is None:
+        protocol = PROTOCOL_VERSION
+
+    json_spec = treespec_to_json(treespec)
+    str_spec = json.dumps((protocol, dataclasses.asdict(json_spec)))
+    return str_spec
+
+def treespec_loads(data: str) -> TreeSpec:
+    protocol, json_schema = json.loads(data)
+
+    if protocol != PROTOCOL_VERSION:
+        raise ValueError(f"Protocol mismatch: {protocol} != {PROTOCOL_VERSION}")
+
```

Comment:
I want to see this formatted like:

```
def treespec_loads(data: str) -> TreeSpec:
  if protocol in _SUPPORTED_PROTOCOLS:
      return _SUPPORTED_PROTOCOLS[protocol].load(data)
  raise ValueError("Unknown protocol: ")
```

So that it is easier for someone to come along in the future. All they have to do is add an additional protocol to this list (and also maybe bump the DEFAULT_PROTOCOL)